File: app/src/main/java/com/geeksville/mesh/util/Utf8ByteLengthFilter.java
/*
 * Copyright (C) 2011 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.geeksville.mesh.util;

import android.text.InputFilter;
import android.text.Spanned;

/**
 * This filter will constrain edits so that the text length is not
 * greater than the specified number of bytes using UTF-8 encoding.
 */
public class Utf8ByteLengthFilter implements InputFilter {
    private final int mMaxBytes;
    public Utf8ByteLengthFilter(int maxBytes) {
        mMaxBytes = maxBytes;
    }
    public CharSequence filter(CharSequence source, int start, int end,
                               Spanned dest, int dstart, int dend) {
        int srcByteCount = 0;
        // count UTF-8 bytes in source substring
        for (int i = start; i < end; i++) {
            char c = source.charAt(i);
            srcByteCount += (c < (char) 0x0080) ? 1 : (c < (char) 0x0800 ? 2 : 3);
        }
        int destLen = dest.length();
        int destByteCount = 0;
        // count UTF-8 bytes in destination excluding replaced section
        for (int i = 0; i < destLen; i++) {
            if (i < dstart || i >= dend) {
                char c = dest.charAt(i);
                destByteCount += (c < (char) 0x0080) ? 1 : (c < (char) 0x0800 ? 2 : 3);
            }
        }
        int keepBytes = mMaxBytes - destByteCount;
        if (keepBytes <= 0) {
            return "";
        } else if (keepBytes >= srcByteCount) {
            return null; // use original dest string
        } else {
            // find end position of largest sequence that fits in keepBytes
            for (int i = start; i < end; i++) {
                char c = source.charAt(i);
                keepBytes -= (c < (char) 0x0080) ? 1 : (c < (char) 0x0800 ? 2 : 3);
                if (keepBytes < 0) {
                    return source.subSequence(start, i);
                }
            }
            // If the entire substring fits, we should have returned null
            // above, so this line should not be reached. If for some
            // reason it is, return null to use the original dest string.
            return null;
        }
    }
}

File: core/common/src/androidMain/kotlin/org/meshtastic/core/common/util/Utf8ByteLengthFilter.kt
/*
 * Copyright (c) 2026 Meshtastic LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.meshtastic.core.common.util

import android.text.InputFilter
import android.text.Spanned

/**
 * An [InputFilter] that constrains text length based on UTF-8 byte count instead of character count. This is
 * particularly useful for fields that must be stored in byte-limited buffers, such as hardware configuration fields.
 *
 * @param maxBytes The maximum allowed length in UTF-8 bytes.
 */
class Utf8ByteLengthFilter(private val maxBytes: Int) : InputFilter {

    private companion object {
        const val ONE_BYTE_LIMIT = '\u0080'
        const val TWO_BYTE_LIMIT = '\u0800'
        const val BYTES_1 = 1
        const val BYTES_2 = 2
        const val BYTES_3 = 3
    }

    override fun filter(
        source: CharSequence,
        start: Int,
        end: Int,
        dest: Spanned,
        dstart: Int,
        dend: Int,
    ): CharSequence? {
        val srcByteCount = countUtf8Bytes(source, start, end)

        // Calculate bytes in dest excluding the range being replaced
        val destLen = dest.length
        var destByteCount = 0
        destByteCount += countUtf8Bytes(dest, 0, dstart)
        destByteCount += countUtf8Bytes(dest, dend, destLen)

        var keepBytes = maxBytes - destByteCount
        return when {
            keepBytes <= 0 -> ""

            keepBytes >= srcByteCount -> null

            else -> {
                for (i in start until end) {
                    val c = source[i]
                    keepBytes -= getByteCount(c)
                    if (keepBytes < 0) {
                        return source.subSequence(start, i)
                    }
                }
                null
            }
        }
    }

    private fun countUtf8Bytes(seq: CharSequence, start: Int, end: Int): Int {
        var count = 0
        for (i in start until end) {
            count += getByteCount(seq[i])
        }
        return count
    }

    private fun getByteCount(c: Char): Int = when {
        c < ONE_BYTE_LIMIT -> BYTES_1
        c < TWO_BYTE_LIMIT -> BYTES_2
        else -> BYTES_3
    }
}
